### PR TITLE
Split PacketHandler up into separate handlers

### PR DESCRIPTION
**Pull request by BomBardyGamer**

### CODE REVIEW

**Review by knightzmc — APPROVED** [review]
there's probably a bit more i could've said but this is fine

File: server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        is PacketInLoginStart -> handleLoginStart(packet)
+        is PacketInEncryptionResponse -> handleEncryptionResponse(packet)
+        else -> Unit
+    }
+
+    private fun handleLoginStart(packet: PacketInLoginStart) {
+        session.player = KryptonPlayer(packet.name, server, session, session.channel.remoteAddress() as InetSocketAddress)
+
+        if (!server.isOnline) {
+            val offlineUUID = UUID.nameUUIDFromBytes("OfflinePlayer:${packet.name}".encodeToByteArray())
```
[the remark this answers is not in the document]
Comment (BomBardyGamer) [reply server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt:51]: Offline UUIDs use UUID v3, which is generated using a namespace (in this case, "OfflinePlayer"), and a key (in this case, the player's name). This was also taken from vanilla.

File: server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    private fun handleEncryptionResponse(packet: PacketInEncryptionResponse) {
+        verifyToken(verifyToken, packet.verifyToken)
+
+        val sharedSecret = server.encryption.decrypt(packet.secret)
+        val secretKey = SecretKeySpec(sharedSecret, "AES")
+        enableEncryption(secretKey)
+
+        Thread({
```
[the remark this answers is not in the document]
Comment (BomBardyGamer) [reply server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt:70]: Yeah that was nicked straight out of vanilla haha, I'll make it use the IO dispatcher.

File: server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                return@Thread
+            }
+            enableCompression()
+
+            session.player.uuid = session.profile.uuid
+            sessionManager.beginPlayState(session)
+        }, "User Authenticator #${UNIQUE_THREAD_ID.incrementAndGet()}")
+            .apply { uncaughtExceptionHandler = DefaultUncaughtExceptionHandler(LOGGER) }
+            .start()
+//        try {
```
[the remark this answers is not in the document]
Comment (BomBardyGamer) [reply server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt:85]: Aye aye captain

File: server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+//        enableCompression()
+//
+//        session.player.uuid = session.profile.uuid
+//        sessionManager.beginPlayState(session)
+    }
+
+    private fun verifyToken(expected: ByteArray, actual: ByteArray) {
+        val decryptedActual = server.encryption.decrypt(actual)
+        require(decryptedActual.contentEquals(expected)) {
+            LOGGER.warn("${session.player.name} failed verification! Expected ${expected.contentToString()}, received ${decryptedActual.contentToString()}")
```
[the remark this answers is not in the document]
Comment (BomBardyGamer) [reply server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt:101]: Yeah good point, I'll make it an error

File: server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    private fun enableCompression() {
+        val threshold = server.config.server.compressionThreshold
+        val compressor = session.channel.pipeline()[PacketCompressor.NETTY_NAME]
+        val decompressor = session.channel.pipeline()[PacketDecompressor.NETTY_NAME]
+
+        if (threshold > 0) {
+            session.sendPacket(PacketOutSetCompression(threshold))
+            if (decompressor is PacketDecompressor) {
+                decompressor.threshold = threshold
+            } else {
```
[the remark this answers is not in the document]
Comment (BomBardyGamer) [reply server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt:130]: Blame vanilla for that one, that is a good point though

[the remark this answers is not in the document]
Comment (BomBardyGamer) [reply server/src/main/kotlin/org/kryptonmc/krypton/packet/handlers/LoginHandler.kt:130]: Also, that will trigger when there isn't a compressor/decompressor in the pipeline, hence the code that adds it present within the `else`

**Pull request merged**